### DISCUSSION TITLE
Understanding validation results

### DISCUSSION THREAD

**Discussion by Jakob-98** [ultralytics/yolov5#8416] [Q&A]
Hi, 

I'm having difficulty explaining the difference in my validation results for a validation set which I first letterbox to 224x224, and a validation set where I use the --size 224 flag. 

**Scenario 1:** I train a model and run 
`python ./val.py --weights ./trained/islands_full_10_90epochs.pt --img 224 --data islands64.yaml --device 0`

My validation dataset is raw/unfiltered images with varying resolution. e.g.: 
![image](https://user-images.githubusercontent.com/37184788/176719257-b7f292f2-a9b5-469d-85c9-e18727f6018a.png)

I assumed this would use the `letterbox` function to resize the images before validating. 

Results (notice inference time): 
![image](https://user-images.githubusercontent.com/37184788/176719435-3554a3fa-aee1-4a1f-9f8a-5bf55344eb6b.png)

**Scenario 2:** I run `python ./val.py --weights ./trained/islands_full_10_90epochs.pt --img 224 --data islands64.yaml --device 0` again, but this time I first letterbox the images to reshape them to 224x224. I do this with the following method, which should be similar to how yolov5 does it:

![image](https://user-images.githubusercontent.com/37184788/176719723-41323b8e-de34-457f-9525-a0f699fc7224.png)

Reference image after letterboxing: 
 
![image](https://user-images.githubusercontent.com/37184788/176719839-5ec1b3a6-cd3c-4e70-b6b3-c3fd7a4f09a2.png)

Results: 
![image](https://user-images.githubusercontent.com/37184788/176719900-141d3cbf-75dc-4121-b377-f45f2e75c51b.png)
Notice the results vary widely. It seems like the validation set is not letterboxed in this case? If that is true, I would expect the inference time to be less since the images would be faster to process (?). 

What explains the differences between these results?

**Comment by glenn-jocher** [comment]
@Jakob-98 thanks for your question! Unfortunately we don't have resources to help analyze and debug custom code/implementations, but I can say that pre-processing times displayed by val.py do not include dataloader time. 

If you want an accurate measure of pre-processing speeds then PyTorch Hub inference can do this. See PyTorch Hub tutorial below for details. In terms of lower metrics, the change would naturally be due to your own code which is beyond the scope of our support.

## YOLOv5 Tutorials

* [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://github.com/ultralytics/yolov5/wiki/Tips-for-Best-Training-Results)&nbsp; ☘️  RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Roboflow for Datasets, Labeling, and Active Learning](https://github.com/ultralytics/yolov5/issues/4975)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://github.com/ultralytics/yolov5/issues/475)
* [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)&nbsp; ⭐ NEW
* [TFLite, ONNX, CoreML, TensorRT Export](https://github.com/ultralytics/yolov5/issues/251) 🚀
* [Test-Time Augmentation (TTA)](https://github.com/ultralytics/yolov5/issues/303)
* [Model Ensembling](https://github.com/ultralytics/yolov5/issues/318)
* [Model Pruning/Sparsity](https://github.com/ultralytics/yolov5/issues/304)
* [Hyperparameter Evolution](https://github.com/ultralytics/yolov5/issues/607)
* [Transfer Learning with Frozen Layers](https://github.com/ultralytics/yolov5/issues/1314)&nbsp; ⭐ NEW
* [Architecture Summary](https://github.com/ultralytics/yolov5/issues/6998)&nbsp; ⭐ NEW

Good luck 🍀 and let us know if you have any other questions!

**Reply by Jakob-98** [reply]
Thank you for the swift reply @glenn-jocher .

Two things are still unclear, which are not particular to my use case/implementation but rather to the validation in general. 
1. If I run the default `./val.py` with --img 224 on e.g. a 2000x2000 image, will the image be resized to 224x224 prior to inference on the model?
2. If **yes** first resizing to 224x224 using the implementation of `letterbox` found in the source should yield the same results, but it doesn't - that is an unexpected/strange result.  If **no**- it is strange that inference time (not preprocessing time) is the same- but in that case I'll figure out why and add it to the discussion.

**Reply by glenn-jocher** [reply]
1. yes
2. as I said before pre-processing times displayed by val.py are not accurate, you can use PyTorch Hub for accurate pre-processing times.

**Reply by Jakob-98** [reply]
Thank you again for your reply :)

It is unusual that first preprocessing to 224x224 _outside_ of the build-in validation pipeline would yield different accuracy/prediction results than processing to 224x224 _inside_ of the built-in validation pipeline. That does not make sense to me. I'll double check tomorrow if this is not PEBCAK on my end, and update the discussion accordingly. 

Regarding preprocessing, that wasn't really my question. Preprocessing times are not really interesting to me, rather whether or not inference time changes depending on resolution... Thank you anyway

I'll update on the first issue tomorrow.

**Reply by glenn-jocher** [reply]
@Jakob-98 inference time is identical no matter what the original image size is since the model never sees the original image size, it only sees the letterboxed input at --imgsz

Preprocessing time may vary, but this is not reflected in val.py

**Reply by Jakob-98** [reply]
I just did another test run- 

I ran a model on my validation set without preprocessing the images beforehand - meaning ~3000x2000 images - with the --image 64 flag. The results were vastly different from first resizing the images to 64x64 beforehand.

I notice that the ./val pipeline does not resize to 64x64, but rather to 64xDIM depending on the original image dimension...

The only other difference is that the bounding box label (class label) is more forgiving in the 64x64 preprocessed image, but this should not influence P/R (?right). 

I do not understand how these results can differ if the validation pipeline indeed first preprocesses the images... perhaps it is a problem on my end. I'll dig some more, but I guess at this point the discussion can be closed
